### ISSUE TITLE
feat: bad-crop tagging and inference-run persistence

### ISSUE THREAD

**Issue by shaoster** [shaoster/glaze#421]
Vision: [Pottery Computer Vision Pipeline](https://gist.github.com/shaoster/a9f2437d570ea884f23c8854d0db5e03)

## Problem / Motivation

Glaze's automatic subject-segmentation pipeline (the crop service backing `piece_image_crop_service`, with a local fallback) sometimes produces visibly wrong crops — over-cropping through a translucent glaze edge, cutting off a handle, latching onto the background, etc. Today there is no way for an admin or piece owner to flag a bad result, and we keep no record of which backend produced a given crop, so we cannot:

- accumulate a corpus of real-world failures to drive model improvement;
- correlate failures with a specific backend/model version once we start
  swapping backends;
- attach future ground-truth annotations to the exact inference that went
  wrong.

This issue is the foundational data layer for milestone
[#2 — Custom Pottery Crop Model — Training Pipeline & Failure-Driven Improvement](https://github.com/shaoster/glaze/milestone/2).
Every later sub-issue (mask-level failure analysis, supervised annotation,
training dataset export, A/B routing, etc.) depends on the artifacts produced
here.

## Proposed Solution

Add two new Django models and wire them into the existing crop pipeline, plus
a minimal "report bad crop" affordance for admins and piece owners and an
admin list view for the resulting reports.

### 1. `CropRun` — one row per inference

Persisted unconditionally by `detect_subject_crop` (in `api/tasks.py`) for
every successful or attempted inference, regardless of whether the resulting
crop was actually applied to the `Piece` / `PieceStateImage`.

Suggested fields (final names settled in implementation):

- `id` (UUID, pk)
- `image` — FK to `Image` (the source image that was cropped)
- `backend_version` — `CharField`, identifies the inference backend
  (e.g. `"rembg-u2net@modal"`, `"rembg-u2netp@local"`). Plain string for now;
  swapped for a backend-registry FK in sub-issue #8.
- `mask_asset` — optional reference to a stored segmentation mask asset
  (Cloudinary public id + cloud name, mirroring `Image`). Nullable so we can
  ship even before the crop service starts returning a raw mask; later
  sub-issues (#3, #8) require it to be populated.
- `crop` — the relative `{x, y, width, height}` JSON the service returned
  (same shape as `Piece.thumbnail_crop` / `PieceStateImage.crop`).
- `latency_ms` — integer wall-clock latency of the inference call.
- `created` — `DateTimeField(auto_now_add=True)`.
- `async_task` — optional FK to `AsyncTask` (the orchestrating
  `detect_subject_crop` invocation), for traceability.
- `status` — `TextChoices` (`success`, `no_subject`, `error`) so we also
  record runs that produced no usable crop.
- `error` — optional text, populated on `error`.

`Meta.ordering = ["-created"]`; index on `(image, -created)` so we can fetch
"latest run for this image" cheaply.

### 2. `CropFailureReport` — one row per flagged bad crop

Created when an admin or piece owner clicks "report bad crop" on a piece's
thumbnail or a `PieceStateImage`.

Suggested fields:

- `id` (UUID, pk)
- `crop_run` — FK to `CropRun` (the inference being flagged). Required so
  every report links to a concrete artifact.
- `reporter` — FK to `settings.AUTH_USER_MODEL`.
- `category` — `TextChoices`. Initial values (open to refinement during
  implementation):
  - `over_cropped` — subject is clipped
  - `under_cropped` — too much background included
  - `wrong_subject` — model latched onto background or wrong object
  - `multi_subject` — multiple pieces visible, wrong one chosen
  - `glaze_edge` — translucent glaze edge mis-segmented
  - `other` — with a free-text note
- `notes` — optional `TextField` for free-form context.
- `created` — `DateTimeField(auto_now_add=True)`.

`Meta.ordering = ["-created"]`.

### 3. Wiring into the existing pipeline

Codebase touchpoints:

- `api/tasks.py:detect_subject_crop` — after the
  `calculate_subject_crop` / `calculate_subject_crop_remote` call, persist a
  `CropRun` (including no-subject and error outcomes). Capture
  `latency_ms` around the inference call. Backend version string derived
  from which branch ran (remote `REMOTE_REMBG_URL` vs. local rembg).
- `tools/piece_image_crop_service.py` — extend the response so the service
  returns the segmentation mask (or a reference to where it was stored) and
  its own backend identifier. Mask storage strategy: upload to Cloudinary
  (same account as `Image`) and return `{cloud_name, cloudinary_public_id}`,
  keeping the existing relative-bbox response for backward compatibility.
  If a full mask-return path is too large to ship in this issue, gate it
  behind a feature flag and leave `CropRun.mask_asset` nullable for now —
  document the follow-up.
- `api/serializers.py` — no behavioral change to the existing
  `detect_subject_crop` dispatches at lines ~473 and ~780; the new
  `CropRun` rows are a side-effect of the task itself.
- New DRF endpoint: `POST /api/crop-failure-reports/` accepting
  `{crop_run, category, notes}`. Permission: authenticated; non-admin users
  may only report runs whose `image` is attached to one of their own
  `Piece`s. Admins may report any run.
- Optional helper endpoint: `GET /api/images/{id}/crop-runs/?latest=1` so
  the UI can resolve "which run produced the currently-displayed crop" when
  rendering the report button.

### 4. UI affordance

In the frontend (touchpoints in `web/src/`), add a "Report bad crop" action
on:

- the piece detail thumbnail, and
- each `PieceStateImage` thumbnail.

Visibility rules:

- piece owner: visible on their own pieces;
- admin (staff): visible on any piece.

Clicking it opens a small dialog (MUI `Dialog`) with the category dropdown
and an optional notes field, then `POST`s to the new endpoint. On success,
show a transient confirmation. Disable / hide the action if no `CropRun`
exists for the image yet (e.g. the crop was set manually). Re-use existing
patterns from `react-conventions` / `react-testing` skills for dialog +
form + async submit.

### 5. Admin

- Register `CropRun` in `api/admin.py` with list columns
  `image`, `backend_version`, `status`, `latency_ms`, `created`, plus
  filters on `backend_version` and `status`.
- Register `CropFailureReport` with list columns `crop_run`, `reporter`,
  `category`, `created`, plus filters on `category` and `reporter`. List
  display should let an admin click through to the underlying `CropRun`
  and the source `Image`.

### 6. Migrations & types

- New Django migration(s) under `api/migrations/` adding the two models.
  Must be wired into `//api:migrations` so they show up in
  `bazel query 'labels(srcs, //api:migrations)'`.
- New DRF serializers + TypeScript types regenerated so
  `web/src/util/types.ts` stays in sync with serializer output (per the
  DoD checklist in `.agents/skills/github-pr/SKILL.md`).

## Acceptance Criteria

- [ ] `CropRun` model exists with the fields listed above (image FK,
  backend_version, mask_asset reference, crop JSON, latency_ms, created,
  status, optional `async_task` FK, optional `error`).
- [ ] `CropFailureReport` model exists with the fields listed above
  (crop_run FK, reporter FK, category, notes, created).
- [ ] `detect_subject_crop` in `api/tasks.py` persists a `CropRun` on every
  invocation — success, no-subject, and error — with measured `latency_ms`
  and a backend-version string derived from the active code path.
- [ ] `tools/piece_image_crop_service.py` returns a mask asset reference
  (or a documented, gated follow-up if mask return is deferred), and the
  task records it on `CropRun.mask_asset` when present.
- [ ] DRF endpoint to create a `CropFailureReport` exists, with permission
  rules: piece owners may only report runs whose source image belongs to
  one of their own pieces; admins may report any run.
- [ ] Frontend exposes a "Report bad crop" action on piece thumbnails and
  `PieceStateImage` thumbnails for the owner and for admins, opening a
  category + notes dialog that posts to the new endpoint.
- [ ] Django admin registers both models with the list views and filters
  described above.
- [ ] Tests cover: `CropRun` creation in all three task outcomes, owner
  vs. non-owner vs. admin permission on the report endpoint, and the
  frontend dialog submit path (per `react-testing` conventions).
- [ ] New migration(s) included in `//api:migrations`
  (`bazel query 'labels(srcs, //api:migrations)'`).
- [ ] Serializer output matches regenerated entries in
  `web/src/util/types.ts`.
- [ ] All linters pass and `rtk bazel test //...` is green.

## Out of Scope

- Building the supervised annotation UI or any ground-truth labeling flow
  (sub-issue #3).
- Replacing rembg with a trained model, or adding a backend registry /
  per-model-version Modal apps (sub-issues #6, #8). `CropRun.backend_version`
  remains a plain string in this issue.
- A/B routing or paired-run comparison (sub-issue #9).
- Per-potter padding settings (sub-issue #3); the hardcoded 10% padding in
  `tools/piece_image_crop_service.py` stays untouched here.
- Mask-level rembg failure analysis or success-criteria definition
  (sub-issue #2) — this issue just makes the data collectible.
- Privacy Policy verification for training-data use (sub-issue #5).

## References

- Milestone: [#2 — Custom Pottery Crop Model](https://github.com/shaoster/glaze/milestone/2)
- Vision doc: [Pottery Computer Vision Pipeline](https://gist.github.com/shaoster/a9f2437d570ea884f23c8854d0db5e03)
- Crop service: `tools/piece_image_crop_service.py`
- Async task: `api/tasks.py:detect_subject_crop`
- Serializer dispatches: `api/serializers.py` (search `detect_subject_crop`)
- Existing models: `api/models.py` (`Image`, `Piece`, `PieceStateImage`,
  `AsyncTask`)
- Local fallback / remote dispatch: `api/utils.py:calculate_subject_crop`,
  `calculate_subject_crop_remote`
- Vision alignment: segmentation-first design doc linked above

## Milestone Cross-References

Foundational ticket — no prerequisites within this milestone.

Blocks:
- [#413 — Rembg Failure-Mode Analysis (mask-level)](https://github.com/shaoster/glaze/issues/413)
- [#420 — Supervised Crop Annotation UI + Per-Potter Padding](https://github.com/shaoster/glaze/issues/420)
- [#414 — Label Studio Seed-Labeling Bootstrap](https://github.com/shaoster/glaze/issues/414)
- [#419 — Per-Model-Version Modal Apps + Backend Registry](https://github.com/shaoster/glaze/issues/419)
- [#418 — A/B Routing, Versioning & Comparison Dashboard](https://github.com/shaoster/glaze/issues/418)

Part of milestone [#2 — Custom Pottery Crop Model](https://github.com/shaoster/glaze/milestone/2).

**Comment by shaoster** [comment]
## Agent Implementation Plan

### Summary

Add `CropRun` — a unified model capturing every crop inference (automated or human-submitted) — plus wiring into the existing crop pipeline, a DRF endpoint for human submissions, frontend "Report bad crop" affordance, and Django admin registration.

### Key Design Decisions (settled in planning)

- **No separate `CropFailureReport`** — human corrections are `CropRun` rows with `source.type = "human"` and `submitter` set.
- **`source` is a JSONField** with schema `{type, backend, deployment, version}` — machine-readable provenance; `backend` is an open string, `version` is `null` until #419 lands a registry.
- **Service renamed** — `tools/piece_image_crop_service.py` → `tools/piece_image_segment_service.py`. Its responsibility is narrowed to pure rembg segmentation. It returns only `{"mask": "<b64>"|null}` — no bbox.
- **Bbox derivation** moves to `run_crop_inference()` on the Django side, computed from the mask alpha channel via `alpha.getbbox()`.
- **Mask format**: rembg RGBA output with RGB channels zeroed (alpha = foreground confidence), base64-encoded in the service response.
- **`run_crop_inference()`** in `api/utils.py` is the single instrumentation point: calls the segment service, derives bbox from alpha, uploads mask to Cloudinary, creates and returns the `CropRun`.

### Rename scope

| Old | New |
|---|---|
| `tools/piece_image_crop_service.py` | `tools/piece_image_segment_service.py` |
| Modal app label `crop` | `segment` |
| `calculate_subject_crop_remote()` | `calculate_subject_mask_remote()` |
| `calculate_subject_crop()` (local) | `calculate_subject_mask()` |
| `CropRun`, `detect_subject_crop`, `thumbnail_crop` | **unchanged** |

---

### 1. `tools/piece_image_segment_service.py`

Rename the file. Narrow responsibility to pure rembg: after rembg produces RGBA output, zero RGB channels and base64-encode:

```python
r, g, b, a = img.split()
mask_img = Image.merge("RGBA", (Image.new("L", img.size, 0), ..., a))
return {"mask": base64.b64encode(buf.getvalue()).decode()}
```

On no-subject: `{"mask": null}`. Remove the existing bbox computation entirely.

### 2. `api/models.py` — `CropRun`

Add after `AsyncTask`:

```python
class CropRun(models.Model):
    class Status(models.TextChoices):
        SUCCESS = "success"; NO_SUBJECT = "no_subject"; ERROR = "error"

    id         = UUIDField(pk)
    image      = FK(Image, PROTECT, related_name="crop_runs")
    source     = JSONField()   # {type, backend, deployment, version}
    submitter  = FK(AUTH_USER_MODEL, null=True, SET_NULL)
    crop       = JSONField(null=True)      # {x,y,width,height} derived from mask alpha
    mask_asset = JSONField(null=True)      # {cloud_name, cloudinary_public_id}
    latency_ms = PositiveIntegerField(null=True)
    status     = CharField(choices=Status)
    error      = TextField(null=True, blank=True)
    notes      = TextField(blank=True)
    async_task = FK(AsyncTask, null=True, SET_NULL)
    created    = DateTimeField(auto_now_add=True)

    class Meta:
        ordering = ["-created"]
        indexes  = [Index(fields=["image", "-created"])]
```

### 3. Migration — `api/migrations/0014_croprun.py`

### 4. `api/utils.py`

Rename `calculate_subject_crop_remote` → `calculate_subject_mask_remote` (now returns `{"mask": ...}`).
Rename `calculate_subject_crop` → `calculate_subject_mask`.

New `run_crop_inference(image, async_task=None) -> CropRun`:
1. Calls `calculate_subject_mask_remote`
2. Decodes b64 mask, derives bbox: `alpha.getbbox()` → normalized `{x,y,width,height}`
3. Uploads mask bytes to Cloudinary via `upload_mask_to_cloudinary()`
4. Creates and returns `CropRun` (success/no_subject/error)

New `upload_mask_to_cloudinary(mask_bytes, image) -> dict`.

### 5. `api/tasks.py`

Replace direct remote call with `run_crop_inference(image, async_task=task)`.

### 6. `api/serializers.py`

`CropRunSerializer` (read) and `CropRunCreateSerializer` (write for human submissions).

### 7. `api/views.py` + `backend/urls.py`

`CropRunViewSet` (create + list) with owner/admin permission gating.
`ImageCropRunsView` (GET /api/images/{id}/crop-runs/).

**[SECURITY]** Non-admin users may only submit/list runs for images on their own pieces.

### 8. `api/admin.py`

Register `CropRun` with `source_type` and `source_backend` display methods.

### 9. Frontend

Regenerate types → add `getImageCropRuns` + `createHumanCropRun` to `api.ts` → new `ReportBadCropDialog.tsx` → wire into `PieceCard` and `PiecePhotoGalleryGrid`.

### 10. Tests

Backend: CropRun creation on success/no-subject/error; owner/non-owner/admin on `POST /api/crop-runs/`.
Frontend: `ReportBadCropDialog` renders, submits, shows confirmation, shows error.
